core/src/main/scala/kafka/utils/NetworkClientBlockingOps.scala

```diff
@@ -0,0 +1,102 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package kafka.utils
+
+import java.io.IOException
+import org.apache.kafka.clients.{ConnectionState, ClientRequest, ClientResponse, NetworkClient}
+import org.apache.kafka.common.Node
+
+import scala.annotation.tailrec
+import scala.collection.JavaConverters._
+
+import org.apache.kafka.common.utils.{Time => JTime}
+
+object NetworkClientBlockingOps {
+  implicit def networkClientBlockingOps(client: NetworkClient): NetworkClientBlockingOps =
+    new NetworkClientBlockingOps(client)
+}
+
+class NetworkClientBlockingOps(val client: NetworkClient) extends AnyVal {
+
+  def blockingReady(node: Node, timeout: Long, retryBackoff: Option[Long])(implicit time: JTime): Boolean = {
+    client.ready(node, time.milliseconds()) || pollUntil(timeout) { (_, now) =>
+      if (client.isReady(node, now))
+        true
+      else if (client.connectionFailed(node))
+        throw new IOException(s"Connection to $node failed")
+      else {
+        retryBackoff.foreach(Thread.sleep)
+        false
+      }
+    }
+  }
+
+  def disconnectAndPoll(nodeId: String, pollTimeout: Long)(implicit time: JTime): Boolean = {
+    client.disconnect(nodeId)
+    client.poll(pollTimeout, time.milliseconds())
+    client.connectionState(nodeId) == ConnectionState.DISCONNECTED
+  }
+
+  def blockingSendAndReceive(request: ClientRequest, timeout: Long)(implicit time: JTime): Option[ClientResponse] = {
+    client.send(request)
```

Comment:
Perhaps. I'm not sure though. It's easy enough for the caller to call `blockingReady` if that is desired, right? I didn't do that in `ControllerChannelManager` because the existing code already handles the exception that would be thrown if the initial connection fails:

```
case e: Throwable => // if the send was not successful, reconnect to broker and resend the message
  warn(("Controller %d epoch %d fails to send request %s to broker %s. " +
    "Reconnecting to broker.").format(controllerId, controllerContext.epoch,
      request.toString, toBroker.toString()), e)
  networkClient.disconnectAndPoll(brokerNode.idString, 0)(time)
  connectToBroker()
  isSendSuccessful = false
  // backoff before retrying the connection and send
  CoreUtils.swallowTrace(Thread.sleep(300))
```

The existing code is not the prettiest, but I tried not to make too many changes at this point. I think we should consider redesigning the controller/broker communication to make better use of `NetworkClient` in the future.